docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,259 @@
+[abci-start]: https://github.com/tendermint/abci/commit/1ab3c747182aaa38418258679c667090c2bb1e0d
+
+## Notes
+
+- <a id=note1></a>**Note 1**: The choice to make all ABCI errors protocol-fatal
+  was intended to avoid the risk that recovering an application error could
+  cause application state to diverge.  Divergence can break consensus, so it's
+  essential to avoid it.
+
+  This is a sound principle, but conflates protocol errors with "mechanical"
+  errors such as timeouts, resoures exhaustion, failed connections, and so on.
+  Because the protocol has no way to distinguish these conditions, the only way
+  for an application to report an error is to panic or crash.
+
+  Whether a node is running in the same process as the application or as a
+  separate process, application errors should not be suppressed or hidden.
+  However, it's important to ensure that errors are handled at a consistent and
+  well-defined point in the protocol: Having the application panic or crash
+  rather than reporting an error means the node sees different results
+  depending on whether the application runs in-process or out-of-process, even
+  if the application logic is otherwise identical.
+
+## Appendix: Known Implementations of ABCI Socket Protocol
+
+This is a list of known implementations of the Tendermint custom socket
+protocol. Note that in most cases I have not checked how complete or correct
+these implementations are; these are based on search results and a cursory
+visual inspection.
+
+- Tendermint Core (Go): [client][socket-client], [server][socket-server]
+- Informal Systems [tendermint-rs](https://github.com/informalsystems/tendermint-rs) (Rust): [client](https://github.com/informalsystems/tendermint-rs/blob/master/abci/src/client.rs), [server](https://github.com/informalsystems/tendermint-rs/blob/master/abci/src/server.rs)
+- Tendermint [js-abci](https://github.com/tendermint/js-abci) (JS): [server](https://github.com/tendermint/js-abci/blob/master/src/server.js)
+- [Hotmoka](https://github.com/Hotmoka/hotmoka) ABCI (Java): [server](https://github.com/Hotmoka/hotmoka/blob/master/io-hotmoka-tendermint-abci/src/main/java/io/hotmoka/tendermint_abci/Server.java)
+- [Tower ABCI](https://github.com/penumbra-zone/tower-abci) (Rust): [server](https://github.com/penumbra-zone/tower-abci/blob/main/src/server.rs)
+- [abci-host](https://github.com/datopia/abci-host) (Clojure): [server](https://github.com/datopia/abci-host/blob/master/src/abci/host.clj)
+- [abci_server](https://github.com/KrzysiekJ/abci_server) (Erlang): [server](https://github.com/KrzysiekJ/abci_server/blob/master/src/abci_server.erl)
+- [py-abci](https://github.com/davebryson/py-abci) (Python): [server](https://github.com/davebryson/py-abci/blob/master/src/abci/server.py)
+- [scala-tendermint-server](https://github.com/intechsa/scala-tendermint-server) (Scala): [server](https://github.com/InTechSA/scala-tendermint-server/blob/master/src/main/scala/lu/intech/tendermint/Server.scala)
```

Comment:
foam team has one in haskell (I believe its in production) https://github.com/f-o-a-m/kepler